lang/cs/Org.Apache.REEF.IO.Tests/TestDataCache.cs

```diff
@@ -0,0 +1,388 @@
+            Assert.True(cache.Materialized.IsPresent());
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.Disk, false));
+            Assert.False(cache.DiskDirectory.IsPresent());
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.InMemoryAsStream, false));
+            Assert.False(cache.MemoryStreams.IsPresent());
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.InMemoryMaterialized, false));
+        }
+
+        /// <summary>
+        /// Tests that using the clean cache option returns the right level.
+        /// </summary>
+        [Fact]
+        public void TestCleanCacheWhileLoweringLevelReturnsRightLevel()
+        {
+            var cache = _context.GetDataCache();
+            SubstituteForBasicDataMover();
+            Assert.Equal(CacheLevel.NotLocal, cache.CacheLevel);
+            Assert.Equal(CacheLevel.Disk, cache.Cache(CacheLevel.Disk, true));
+            Assert.Equal(CacheLevel.Disk, cache.CacheLevel);
+            Assert.Equal(CacheLevel.InMemoryAsStream, cache.Cache(CacheLevel.InMemoryAsStream, true));
+            Assert.Equal(CacheLevel.InMemoryAsStream, cache.CacheLevel);
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.InMemoryMaterialized, true));
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.CacheLevel);
+            Assert.Equal(CacheLevel.NotLocal, cache.CleanCacheAtLevel(CacheLevel.InMemoryMaterialized));
+            Assert.Equal(CacheLevel.NotLocal, cache.CacheLevel);
+        }
+
+        /// <summary>
+        /// Tests that cleaning the cache cleans all upper level caches.
+        /// </summary>
+        [Fact]
+        public void TestCleanCacheClearsAllUpperCacheLevels()
+        {
+            var cache = _context.GetDataCache();
+            SubstituteForBasicDataMover();
+            Assert.Equal(CacheLevel.Disk, cache.Cache(CacheLevel.Disk, false));
+            Assert.Equal(CacheLevel.InMemoryAsStream, cache.Cache(CacheLevel.InMemoryAsStream, false));
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.InMemoryMaterialized, true));
+            Assert.Equal(CacheLevel.NotLocal, cache.CleanCacheAtLevel(CacheLevel.InMemoryMaterialized));
+        }
+
+        /// <summary>
+        /// Tests clear cache cleans all state and returns right level.
+        /// </summary>
+        [Fact]
+        public void TestClearCache()
+        {
+            var cache = _context.GetDataCache();
+            SubstituteForBasicDataMover();
+            Assert.Equal(CacheLevel.Disk, cache.Cache(CacheLevel.Disk, false));
+            Assert.Equal(CacheLevel.InMemoryAsStream, cache.Cache(CacheLevel.InMemoryAsStream, false));
+            Assert.Equal(CacheLevel.InMemoryMaterialized, cache.Cache(CacheLevel.InMemoryMaterialized, false));
```

Comment:
Given that this is an object that keeps state, I opt to do it for every test and every state because both the values passed to the object's methods and the order of method invocation/order of state change matters. Every test differs slightly in one (or more) of the above.